tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -0,0 +1,858 @@
+            throw new TypingInputException(
+              s"Parser error in Typing!EmptySeq($elemTypeText): ${e.msg}"
+            )
+        }
+
+      //******************************************** GENERAL OPERATORS **************************************************
+      case OperEx(op, args @ _*) if op == TlaOper.eq || op == TlaOper.ne =>
+        // x = y, x /= y
+        val a = varPool.fresh
+        val opsig = OperT1(Seq(a, a), BoolT1())
+        mkExRefApp(opsig, args)
+
+      case OperEx(TlaOper.apply, NameEx(name), args @ _*) =>
+        // F(e_1, ..., e_n)
+        mkAppByName(ref, name, args.map(this(_)): _*)
+
+      case OperEx(TlaOper.apply, opName, args @ _*) =>
+        throw new TypingException(
+          "Bug in ToEtcExpr. Expected an operator name, found: " + opName
+        )
+
+      case OperEx(
+            TlaOper.chooseBounded,
+            NameEx(bindingVar),
+            bindingSet,
+            pred
+          ) =>
+        // CHOOSE x \in S: P
+        // the principal type of CHOOSE is (a => Bool) => a
+        val a = varPool.fresh
+        val chooseType = OperT1(Seq(OperT1(Seq(a), BoolT1())), a)
+        // CHOOSE implicitly introduces a lambda abstraction: λ x ∈ S. P
+        val chooseLambda =
+          mkAbs(BlameRef(ex.ID), this(pred), (bindingVar, this(bindingSet)))
+        // the resulting expression is (((a => Bool) => a) (λ x ∈ S. P))
+        mkApp(ref, Seq(chooseType), chooseLambda)
+
+      case OperEx(TlaOper.chooseUnbounded, NameEx(bindingVar), pred) =>
+        // CHOOSE x: P
+        // the principal type of CHOOSE is (a => Bool) => a
+        //
+        // Igor: I am not sure that this is a good solution, as the type checker will not propagate the possible
+        // values of b from P. This expression will always give us a polytype.
+        val a = varPool.fresh
+        val chooseType = OperT1(Seq(OperT1(Seq(a), BoolT1())), a)
+        // unbounded CHOOSE implicitly introduces a lambda abstraction: λ x ∈ Set(b). P
+        val b = varPool.fresh
+        val chooseLambda = mkAbs(
+          BlameRef(ex.ID),
+          this(pred),
+          (bindingVar, mkConst(BlameRef(ex.ID), SetT1(b)))
+        )
+        // the resulting expression is (((a => Bool) => a) (λ x ∈ Set(b). P))
+        mkApp(ref, Seq(chooseType), chooseLambda)
+
+      //******************************************** LET-IN ****************************************************
+      case LetInEx(body, declarations @ _*) =>
+        declarations.foldRight(this(body)) {
+          case (decl, term) => this(decl, term)
+        }
+
+      //******************************************** BOOLEANS **************************************************
+      case OperEx(op, args @ _*)
+          if op == TlaBoolOper.and || op == TlaBoolOper.or || op == TlaBoolOper.equiv || op == TlaBoolOper.implies =>
+        // A /\ B, A \/ B, A <=> B, A => B
+        val nBools = List.fill(args.length)(BoolT1())
```

Comment:
Technically, `equiv` and `implies` admit only 2 arguments. Perhaps the translator should complain immediately, if this is not the case.

Reply:
Yeah, let's do catch that. I'll separate those out.